### DISCUSSION TITLE
Add event hooks to execute code (like forwarding logs to an external service)

### DISCUSSION THREAD

**Discussion by 0xForerunner** [tokio-rs/tracing#2669] [Q&A] [answered]
Perhaps there is already a way to do this that I'm not aware of. Basically, what I would like to do is have a custom layer that can send off http requests to an external logging service like paper trail. As far as I can tell, the only way to accomplish this currently is to hack `format_event` to send the log through an mpsc channel to a tokio::task executes the async request. It would be very nice if there were a better way to do this. Event hooks. could be added to layers to solve this problem!

**Comment by davidbarsky** [comment, marked as answer]
`tracing` generally assumes that methods on layers are synchronous, and therefore, free to call from an async context without worrying about blocking. While you are _technically_ correct about changing `format_event` to send logs through an MSPC channel to a Tokio task, we offer two, substantially more paved paths to tackle this issue:

- Implement your own [`MakeWriter`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/trait.MakeWriter.html#examples) that sends the formatted message over an MSPC channel. Note that `tracing-appender` is _basically_ this, except it sends logs to a background _thread_, not a task.
- If `MakeWriter` isn't sufficient for whatever reason, writing your own layer is a decently lightweight option! For a guide, I'd take a look at this blog post: https://burgers.io/custom-logging-in-rust-using-tracing/.

That being said, I don't really recommend sending logs from within the application if you can avoid it—I think you're significantly better off writing logs to `stdout`/`stderr` and having some external logging daemon forward the logs to a service.

**Reply by 0xForerunner** [reply]
Hey sorry I never replied to you on this. Appreciate the suggestions! They both seem good. Sending the logs directly I think is the only viable solution for me. I'd like to not lose the log level information when sending the logs to paper trail. I'll give the MakeWriter a shot first and see how that goes. Thanks!

**Reply by 0xForerunner** [reply]
@davidbarsky So after some investigation I've come to the conclusion that there is no _good_ way to do this sadly. To send my logs to the remote service I require both the formatted logs *and* the log level. It appears the only way to  access to both of these is to hijack `format_event` method, which seems pretty hacky. I think this solution will work fine, even if not ideal.

I think the root cause for this difficulty is that `format_event` is actually doing 2 jobs, when it should probably only being doing one. It's both formatting the event *and* writing  to the writer, so there is no easy way to inject logic in between those two steps.
